logstash-core/src/main/java/org/logstash/settings/Boolean.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.settings;
+
+public class Boolean extends Coercible<java.lang.Boolean> {
+
+    public Boolean(String name, boolean defaultValue) {
+        super(name, defaultValue, true, noValidator());
+    }
+
+    public Boolean(String name, boolean defaultValue, boolean strict) {
+        super(name, defaultValue, strict, noValidator());
+    }
+
+    @Override
+    public java.lang.Boolean coerce(Object obj) {
+        if (obj instanceof String) {
+            switch((String) obj) {
+                case "true": return true;
+                case "false": return false;
+                default: throw new IllegalArgumentException("could not coerce " + value() + " into a boolean");
+            }
+        }
+        if (obj instanceof java.lang.Boolean) {
+            return (java.lang.Boolean) obj;
+        }
+        throw new IllegalArgumentException("could not coerce " + value() + " into a boolean");
+    }
```

Comment:
When we fail to coerce, our message contains the _existing_ value, not the one we attempted to coerce:

> ~~~
> ╭─{ rye@perhaps:~/src/elastic/logstash@main (validate/feature/reimplement_config_classes_in_java ✘) }
> ╰─● (echo 'api.enabled: bananas' > config/logstash.yml) && bin/logstash --log.level=debug -e 'input { generator { count => 1 } }'
> Using system java: /Users/rye/.jenv/shims/java
> ERROR: Failed to load settings file from "path.settings". Aborting... path.setting=/Users/rye/src/elastic/logstash@main/config, exception=Java::JavaLang::IllegalArgumentException, message=>could not coerce true into a boolean
> [FATAL] 2024-09-10 14:03:47.613 [main] Logstash - Logstash stopped processing because of an error: (SystemExit) exit
> org.jruby.exceptions.SystemExit: (SystemExit) exit
>         at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:921) ~[jruby.jar:?]
>         at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:880) ~[jruby.jar:?]
>         at Users.rye.src.elastic.logstash_at_40_main.lib.bootstrap.environment.<main>(/Users/rye/src/elastic/logstash@main/lib/bootstrap/environment.rb:90) ~[?:?]
> [error: 1 (00:00:05)]                                                                                                                                                                                                                                                                                                                                                                                                            
> 
> ~~~

```suggestion
    public java.lang.Boolean coerce(Object obj) {
        if (obj instanceof String) {
            switch((String) obj) {
                case "true": return true;
                case "false": return false;
                default: throw new IllegalArgumentException(coercionFailureMessage(obj));
            }
        }
        if (obj instanceof java.lang.Boolean) {
            return (java.lang.Boolean) obj;
        }
        throw new IllegalArgumentException(coercionFailureMessage(obj));
    }

    private String coercionFailureMessage(Object obj) {
        return String.format("Cannot coerce `%s` to boolean", obj);
    }
```

It would also likely be helpful if the crashing exception included _which_ setting we were attempting coerce a value for:

```suggestion
    public java.lang.Boolean coerce(Object obj) {
        if (obj instanceof String) {
            switch((String) obj) {
                case "true": return true;
                case "false": return false;
                default: throw new IllegalArgumentException(coercionFailureMessage(obj));
            }
        }
        if (obj instanceof java.lang.Boolean) {
            return (java.lang.Boolean) obj;
        }
        throw new IllegalArgumentException(coercionFailureMessage(obj));
    }

    private String coercionFailureMessage(Object obj) {
        return String.format("Cannot coerce `%s` to boolean (%s)", obj, this.getName());
    }
```

> ~~~
> ╭─{ rye@perhaps:~/src/elastic/logstash@main (validate/feature/reimplement_config_classes_in_java ✘) }
> ╰─● (echo 'api.enabled: bananas' > config/logstash.yml) && bin/logstash --log.level=debug -e 'input { generator { count => 1 } }'
> Using system java: /Users/rye/.jenv/shims/java
> ERROR: Failed to load settings file from "path.settings". Aborting... path.setting=/Users/rye/src/elastic/logstash@main/config, exception=Java::JavaLang::IllegalArgumentException, message=>Cannot coerce `bananas` to boolean (api.enabled)
> [FATAL] 2024-09-10 14:20:03.107 [main] Logstash - Logstash stopped processing because of an error: (SystemExit) exit
> org.jruby.exceptions.SystemExit: (SystemExit) exit
>         at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:921) ~[jruby.jar:?]
>         at org.jruby.RubyKernel.exit(org/jruby/RubyKernel.java:880) ~[jruby.jar:?]
>         at Users.rye.src.elastic.logstash_at_40_main.lib.bootstrap.environment.<main>(/Users/rye/src/elastic/logstash@main/lib/bootstrap/environment.rb:90) ~[?:?]
> [error: 1 (00:00:05)]                                                                                                                                                                                                                                                                                                                                                                                                            
> 
> ~~~

Reply:
Thank's @yaauie for catching it. I've updated with your suggestion plus covered with a unit test.